File: unixfs/fs-dirent.go
package unixfs

import (
	"io/fs"
	"os"

	"github.com/pkg/errors"
)

// FSDirEntry implements fs.DirEntry with a FSCursorDirent and associated FileInfo.
type FSDirEntry struct {
	// ent is the directory entry
	ent FSCursorDirent
	// fileInfo is the file information
	fileInfo fs.FileInfo
}

// NewFSDirEntry constructs a new directory entry from a unixfs.FSCursorDirent.
//
// fileInfo can be nil but will return ErrInfoUnavailable for
// if fileInfo is nil permissions will be set to defaults.
func NewFSDirEntry(ent FSCursorDirent, fileInfo fs.FileInfo) fs.DirEntry {
	return &FSDirEntry{
		ent:      ent,
		fileInfo: fileInfo,
	}
}

// Name returns the name of the file (or subdirectory) described by the entry.
// This name is only the final element of the path (the base name), not the entire path.
// For example, Name would return "hello.go" not "home/gopher/hello.go".
func (e *FSDirEntry) Name() string {
	return e.ent.GetName()
}

// IsDir reports whether the entry describes a directory.
func (e *FSDirEntry) IsDir() bool {
	return e.ent.GetIsDirectory()
}

// Type returns the type bits for the entry.
// The type bits are a subset of the usual FileMode bits, those returned by the FileMode.Type method.
func (e *FSDirEntry) Type() fs.FileMode {
	var defaultMode fs.FileMode
	if e.fileInfo != nil {
		defaultMode = e.fileInfo.Mode()
	} else {
		if e.ent.GetIsDirectory() {
			defaultMode = fs.ModeDir | 0555
		} else {
			defaultMode = 0444
		}
	}

	typ := NodeTypeToMode(e.ent, defaultMode.Perm())
	if typ == os.ModeIrregular {
		return defaultMode
	}
	return typ
}

// Info returns the FileInfo for the file or subdirectory described by the entry.
// The returned FileInfo may be from the time of the original directory read
// or from the time of the call to Info. If the file has been removed or renamed
// since the directory read, Info may return an error satisfying errors.Is(err, ErrNotExist).
// If the entry denotes a symbolic link, Info reports the information about the link itself,
// not the link's target.
func (e *FSDirEntry) Info() (fs.FileInfo, error) {
	if e.fileInfo == nil {
		return nil, errors.New("file info unavailable")
	}
	return e.fileInfo, nil
}

// _ is a type assertion
var _ fs.DirEntry = (*FSDirEntry)(nil)

File: unixfs/fs-dirent.ts
import type { FSCursorDirent, FSCursorNodeType } from './fs-cursor.js'
import type { FileInfo } from './file-info.js'
import { nodeTypeToMode } from './fs-node-type.js'

// ModeDir is the directory bit in a file mode (matching Go os.ModeDir).
const ModeDir = 0x80000000
// ModeIrregular is the irregular bit in a file mode.
const ModeIrregular = 0x00080000
// ModeTypeMask covers all type bits.
const ModeTypeMask =
  ModeDir |
  0x08000000 |
  ModeIrregular |
  0x04000000 |
  0x02000000 |
  0x01000000 |
  0x00800000 |
  0x00400000 |
  0x00200000 |
  0x00100000

// FSDirEntry implements a directory entry with a FSCursorDirent and associated FileInfo.
export class FSDirEntry {
  private readonly ent: FSCursorDirent
  private readonly fileInfo: FileInfo | null

  constructor(ent: FSCursorDirent, fileInfo: FileInfo | null) {
    this.ent = ent
    this.fileInfo = fileInfo
  }

  // getName returns the name of the file or subdirectory described by the entry.
  // This name is only the final element of the path (the base name).
  getName(): string {
    return this.ent.getName()
  }

  // isDir reports whether the entry describes a directory.
  isDir(): boolean {
    return this.ent.getIsDirectory()
  }

  // getType returns the type bits for the entry.
  getType(): number {
    let defaultMode: number
    if (this.fileInfo) {
      defaultMode = this.fileInfo.getMode()
    } else if (this.ent.getIsDirectory()) {
      defaultMode = ModeDir | 0o555
    } else {
      defaultMode = 0o444
    }

    const typ = nodeTypeToMode(this.ent, defaultMode & 0o777)
    if (typ === ModeIrregular) {
      return defaultMode & ModeTypeMask
    }
    return typ & ModeTypeMask
  }

  // getInfo returns the FileInfo for the file or subdirectory.
  // Throws if file info is unavailable.
  getInfo(): FileInfo {
    if (!this.fileInfo) {
      throw new Error('file info unavailable')
    }
    return this.fileInfo
  }
}
